### PR TITLE
R4R: Cherry pick token id validation

### PR DESCRIPTION
**Pull request by zhangyelong**
Resolves: #1873 
Ref: #1874

### CODE REVIEW

File: types/coin.go [hunk 1]
```diff
@@ -58,7 +58,7 @@ func (coin Coin) IsValid() bool {
 	if coin.IsNegative() {
 		return false
 	}
-	if coin.Denom != IrisAtto && !strings.HasSuffix(coin.Denom, MinDenomSuffix) {
+	if coin.Denom != IrisAtto && (!strings.HasSuffix(coin.Denom, MinDenomSuffix) || strings.HasPrefix(coin.Denom, Iris + "-")) {
```
Comment (zhangyelong) [inline types/coin.go]: This will reject `iris-min` etc.

File: types/coin.go [hunk 2]
```diff
@@ -632,6 +634,23 @@ func ParseCoins(coinsStr string) (coins Coins, err error) {
 	return coins, nil
 }
 
+func IsCoinNameValid(coinName string) bool  {
```
Comment (haifengxi) [inline types/coin.go]: Simply implementation as:
`{ return reCoinNameCompiled.MatchString(coinName) }`

File: types/coin.go [hunk 3]
```diff
@@ -632,6 +634,23 @@ func ParseCoins(coinsStr string) (coins Coins, err error) {
...
+	if !reCoinNameCompiled.MatchString(coinName) {
+		return false
+	}
+	return true
+}
+
+func IsCoinMinDenomValid(denom string) bool  {
+	if denom != IrisAtto && !strings.HasSuffix(denom, MinDenomSuffix) {
+		return false
+	}
```
Comment (haifengxi) [inline types/coin.go]: Remaining lines can be simplified as:
`return reDenomCompiled.MatchString(denom)`

File: types/coin.go [hunk 4]
```diff
@@ -632,6 +634,23 @@ func ParseCoins(coinsStr string) (coins Coins, err error) {
...
 
+func IsCoinNameValid(coinName string) bool  {
+	if !reCoinNameCompiled.MatchString(coinName) {
+		return false
+	}
+	return true
+}
+
+func IsCoinMinDenomValid(denom string) bool  {
+	if denom != IrisAtto && !strings.HasSuffix(denom, MinDenomSuffix) {
```
Comment (haifengxi) [inline types/coin.go]: Should check for invalid `iris-min`

File: types/coin.go [hunk 5]
```diff
@@ -58,7 +58,7 @@ func (coin Coin) IsValid() bool {
 	if coin.IsNegative() {
 		return false
 	}
```
Comment (haifengxi) [inline types/coin.go]: Could simply remaining lines as:
`return IsCoinMinDenomValid(coin.Denom)`

File: types/coin_test.go [hunk 6]
```diff
@@ -506,3 +506,55 @@ func TestAmountOf(t *testing.T) {
 		assert.Equal(t, NewInt(tc.amountOfTREE), tc.coins.AmountOf("tree-min"))
 	}
 }
+
+func TestIsCoinNameValid(t *testing.T) {
+
```
Comment (haifengxi) [inline types/coin_test.go]: Add test case for `iris-min`

**Pull request merged**